Support geo-replication for azurerm_container_registry resource

jcorioland opened the pull request:
This PR improve the `azurerm_container_registry` resource provider to support ACR geo-replication.
- possible to enable geo-replication in multiple Azure locations
- possible to add/remove locations on update
- acceptance tests cover enabling geo-replication in multiple locations, updating locations, updating Sku from Premium to Basic (that does not supports geo-replication)

Fixes #2047

jsturtevant commented on azurerm/resource_arm_container_registry.go:
```diff
@@ -57,6 +57,20 @@ func resourceArmContainerRegistry() *schema.Resource {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
```
Is this needed?  Could we calculate this on the fly based on the `georeplication_locations` parameters? This would avoid the case of these two parameters getting out of sync.  See comment below about one of the edge cases.

jcorioland replied:
I've removed the field and use the georeplication_locations field to determine if it is enabled or not.

jsturtevant commented on azurerm/resource_arm_container_registry.go:
```diff
@@ -203,6 +247,10 @@ func resourceArmContainerRegistryUpdate(d *schema.ResourceData, meta interface{}
 
+	if geoReplicationEnabled && strings.ToLower(sku) != strings.ToLower(string(containerregistry.Premium)) {
+		return fmt.Errorf("`georeplication_enabled` can only be enabled for a Premium Sku.")
+	}
```
In the case where the user enables geo-replication then turns it off on the update, this will not currently remove the replications.

jcorioland replied:
update scenario with turning geo-replication off or updating to a lower Sku than premium is now implemented and covered by acceptance tests.

jsturtevant commented on azurerm/resource_arm_container_registry.go:
```diff
@@ -227,8 +311,53 @@ func resourceArmContainerRegistryUpdate(d *schema.ResourceData, meta interface{}
 }
 
+func getReplicationLocationsToDelete(d *schema.ResourceData) []string {
+	oldGeoreplifationLocations, newGeoreplicationLocations := d.GetChange("georeplication_locations")
```
Also, consider during this time if the SKU changes from premium to basic, do you need to remove the replications or does the change in sku automatically drop the replications?

jcorioland replied:
When updating an SKU from Premium to Basic, it automatically deletes all the replications. This is covered by tests.

jsturtevant commented on azurerm/resource_arm_container_registry_test.go:
```diff
@@ -221,6 +222,27 @@ func TestAccAzureRMContainerRegistry_update(t *testing.T) {
+				Config: config,
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMContainerRegistryExists("azurerm_container_registry.test"),
+					testCheckAzureRMContainerRegistryGeoreplicationExists("azurerm_container_registry.test", georeplicationLocations),
```
Consider adding test case for the scenario where the user switches skus and updates to the geo-replication locations

dcaro replied:
Also you may want to consider verifying some values, not just their existence.

dcaro commented on azurerm/resource_arm_container_registry.go:
```diff
@@ -57,6 +57,20 @@ func resourceArmContainerRegistry() *schema.Resource {
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
```
You should use the ValidateFunc to verify the value passed.

jcorioland replied:
It seems that location does not get validation in the whole provider (see https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/location.go). How would you validate that? Having a fixed list of region could introduce a blocker when new regions are available. And potentially, the location could be validated for all different services to make sure it is available in the region.

jcorioland replied:
Done - and ordering does not matter in that case.

jcorioland commented:
@tombuildsstuff I think this PR is ready for being reviewed. Thanks!

jcorioland commented:
@tombuildsstuff thanks for the review ! :) I will do the suggested modifications. I also thought about externalizing the geo-replication into it's own resource. The thing that makes me choose to bundle it with the container registry resource is that it is really dependent on the Sku.

For example, if you create an ACR in Basic Mode, and then want to add a geo replication, you need to update to premium, so you need to 1. Update the ACR with the ACR resource, and 2. update the geo-replication. Same thing if you want to downgrade the Sku, you need to remove the geo-replication first. This is why I chose to bundle everything in the container registry resource. Happy to discuss this point with you if you think it does not make sense at the end.

jcorioland commented on azurerm/resource_arm_container_registry.go:
```diff
@@ -142,6 +151,11 @@ func resourceArmContainerRegistryCreate(d *schema.ResourceData, meta interface{}
 	}
 
+	// if locations have been specified for geo-replication then, the SKU has to be Premium
+	if len(geoReplicationLocations) > 0 && strings.ToLower(sku) != strings.ToLower(string(containerregistry.Premium)) {
```
Updates done.

jcorioland commented:
@tombuildsstuff please let me know if you think there are some remaining blockers to merge this PR. Thanks!

jcorioland commented on azurerm/resource_arm_container_registry.go:
```diff
@@ -142,6 +153,11 @@ func resourceArmContainerRegistryCreate(d *schema.ResourceData, meta interface{}
 	}
 
+	// if locations have been specified for geo-replication then, the SKU has to be Premium
+	if len(geoReplicationLocations) > 0 && !strings.EqualFold(sku, string(containerregistry.Premium)) {
```
Done

jcorioland commented on azurerm/resource_arm_container_registry.go:
```diff
@@ -57,6 +57,16 @@ func resourceArmContainerRegistry() *schema.Resource {
+			"georeplication_locations": {
+				Type:             schema.TypeList,
+				Optional:         true,
+				StateFunc:        azureRMNormalizeLocation,
```
Done

jcorioland commented on azurerm/resource_arm_container_registry.go:
```diff
@@ -227,8 +276,93 @@ func resourceArmContainerRegistryUpdate(d *schema.ResourceData, meta interface{}
+	for _, locationToCreate := range replicationLocationsToCreate {
+		replication := containerregistry.Replication{
+			Location: &locationToCreate,
+			Name:     &locationToCreate,
```
I wanted to keep it simple: not expose this property and default it to the location, as it is done by the Azure API.

katbyte replied:
👍

jcorioland commented on azurerm/resource_arm_container_registry.go:
```diff
@@ -227,8 +276,93 @@ func resourceArmContainerRegistryUpdate(d *schema.ResourceData, meta interface{}
+	ctx := meta.(*ArmClient).StopContext
+	log.Printf("[INFO] preparing to apply geo-replications for AzureRM Container Registry.")
+
+	replicationLocationsToCreate := getReplicationLocationsToCreate(oldGeoReplicationLocations, newGeoReplicationLocations)
```
I am not sure about what should be pulled out?

katbyte replied:
This function isn't used anywhere else so i think its code should be put here.

Also, i think this could all be vastly simplified:
```go
createLocations := make(map[string]bool)
	for _, nl := range newLocations {
		createLocations[nl] = false
	}
	for _, ol := range oldLocations {
		if _, ok := createLocations[ol]; ok {
			createLocations[ol] = true
		}
	}
	
	for _, locationToCreate := range createLocations {
		if ! createLocations[locationToCreate]  {
			continue
		}
		//create location
	}

	// delete not needed geo-replication locations
	for _, locationToDelete := range replicationLocationsToDelete {
		if _, ok := createLocations[locationToDelete]; !ok  {
			continue
		}
		// delete locatio
	}
```

what do you think?

jcorioland replied:
Hey @katbyte - I've been able to refactor the update as you've suggested. But I've got a difficulty with using a TypeSet for location, so I rollback to TypeList. 

When using a Set, and updating the location from "eastus","westus" to "eastus","westeurope", the eastus region is not considered as a new value (which make sense in a way) but this behavior does not allow to do a diff between old and new locations:
![image](https://user-images.githubusercontent.com/4435138/48268448-49f51c00-e435-11e8-8a8f-435efe321811.png)

When using a list, I don't have these issue and I am able to diff:

![image](https://user-images.githubusercontent.com/4435138/48268472-5da08280-e435-11e8-9d71-1f6f0c372ee9.png)

I am relatively new to Terraform, so maybe there is another option I can explore?

jcorioland commented on azurerm/resource_arm_container_registry.go:
```diff
@@ -283,6 +419,27 @@ func resourceArmContainerRegistryRead(d *schema.ResourceData, meta interface{})
+
+		for i, value := range replicationValues {
+			valueLocation := azureRMNormalizeLocation(*value.Location)
+			if valueLocation != azureRMNormalizeLocation(*location) {
```
Done

jcorioland commented on azurerm/resource_arm_container_registry.go:
```diff
@@ -283,6 +419,27 @@ func resourceArmContainerRegistryRead(d *schema.ResourceData, meta interface{})
+		georeplication_locations := make(map[int]interface{}, len(replicationValues))
+
+		for i, value := range replicationValues {
+			valueLocation := azureRMNormalizeLocation(*value.Location)
```
Done

jcorioland commented on azurerm/resource_arm_container_registry_test.go:
```diff
@@ -221,6 +223,54 @@ func TestAccAzureRMContainerRegistry_update(t *testing.T) {
+				Config: updatedConfig,
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMContainerRegistryExists("azurerm_container_registry.test"),
+					testCheckAzureRMContainerRegistryGeoreplications("azurerm_container_registry.test", "Premium", nil),
```
Same comment than above

jcorioland commented on azurerm/resource_arm_container_registry.go:
```diff
@@ -227,8 +276,93 @@ func resourceArmContainerRegistryUpdate(d *schema.ResourceData, meta interface{}
+		}
+
+		err = future.WaitForCompletionRef(ctx, replicationClient.Client)
+		if err != nil {
```
Done

jcorioland commented on azurerm/resource_arm_container_registry.go:
```diff
@@ -227,8 +276,93 @@ func resourceArmContainerRegistryUpdate(d *schema.ResourceData, meta interface{}
+		}
+
+		err = future.WaitForCompletionRef(ctx, replicationClient.Client)
+		if err != nil {
```
Done

jcorioland commented on azurerm/resource_arm_container_registry.go:
```diff
@@ -251,7 +385,9 @@ func resourceArmContainerRegistryRead(d *schema.ResourceData, meta interface{})
-	if location := resp.Location; location != nil {
+
+	location := resp.Location
+	if location != nil {
```
I added tests != nil when deferencing location below

jcorioland commented on azurerm/resource_arm_container_registry.go:
```diff
@@ -227,8 +276,93 @@ func resourceArmContainerRegistryUpdate(d *schema.ResourceData, meta interface{}
+	log.Printf("[INFO] preparing to apply geo-replications for AzureRM Container Registry.")
+
+	replicationLocationsToCreate := getReplicationLocationsToCreate(oldGeoReplicationLocations, newGeoReplicationLocations)
+	replicationLocationsToDelete := getReplicationLocationsToDelete(oldGeoReplicationLocations, newGeoReplicationLocations)
```
Done, see comment above.

jcorioland commented:
Hi @katbyte,

I just ran the tests for the ACR resource on my environment and they do not fail. Can you please be more specific on what is wrong with the list and causes panic?
About my issue when using a set, it's just that I miss the information that `westus` should be removed.

In other words, regarding the Set below, how can I differentiate that `eastus` shoud be ignored and `westus` should be removed?
![image](https://user-images.githubusercontent.com/4435138/48612386-2c6e0800-e989-11e8-80b0-27dbe04aeace.png)

What am I missing?
Thanks

jcorioland commented:
@katbyte, unfortunately no for the region. In the test I update locations from eastus,westus to eastus,westeurope so only westus should be removed. Not eastus. But when using a Set I don't received the information compared to using a List...

katbyte commented:
@jcorioland,

Could you write a test exhibiting this behaviour so I can see it myself and try and fix it? If that is the case, I would suggest using `d.HasChange`, and then if that is true get the entire new set with `d.Get` instead of just the change.

jcorioland commented:
@katbyte yes will do that by the end of week. Thanks

jcorioland commented:
Hi @katbyte - I've just created a new branch, with a fake resource and a fake unit test to illustrate the problem I have replacing List by a Set for the locations: https://github.com/jcorioland/terraform-provider-azurerm/commit/46f2a467eda8d7d34e4c7dfa2ffb9d8be6234ed1

The `azurerm_issue_use_set` resource has a simple schema with two properties:
- locations_list, a list of locations
- locations_set, a set of locations

The unit test is creating the fake resource with this config:

```
resource "azurerm_issue_use_set" "test" {
  locations_list      = ["eastus", "westeurope"]
  locations_set		  = ["eastus", "westeurope"]
}
```

The create opration only loops on the set/list values and display the content:

```
########## CREATE LIST ##########
Location (list): eastus
Location (list): westus
########## CREATE SET ##########
Location (set): westus
Location (set): eastus
```

Then, the test updates the config:

```
resource "azurerm_issue_use_set" "test" {
  locations_list      = ["eastus", "westeurope"]
  locations_set		  = ["eastus", "westeurope"]
}
```

The update function also just display the content of the set and the list, using both d.Get and d.GetChange:

```
########## UPDATE LIST ##########
locations_list variable has change
Location (list): eastus
Location (list): westeurope
Old location (list): eastus
Old location (list): westus
New location (list): eastus
New location (list): westeurope
########## UPDATE SET ##########
locations_set variable has change
Location (set): westeurope
Old location (set): westus
Old location (set): eastus
New location (set): westeurope
```

For the List variable, the d.Get returns eastus and westeurope
For the Set variable, the d.Get returns only westeurope
For the List variable, the d.GetChange returns old = eastus,westus - new = eastus, westeurope
For the Set variable, the d.GetChange returns old = eastus,westus - new = westeurope
For Both List & Set, the d.HasChange returns true.

With the List type, I am able to diff the old and new location and know that I need to keep eastus but need to delete westus.
With the Set type, I am not able to do it. The only information I have is that I need to add westeurope, but I am not able to determine what I need to do with the two old locations (or I am really missing something ^^)

Hope this unit tests will help to understand the issue I am facing :)

Thanks!
Julien

katbyte commented:
Thank you for the very explicit unit test @jcorioland, much appreciated as it made tracking this down pretty quick.

It seems that the issue is not with `d.*` calls, but the schema diff/state functions. By changing it to this:

```go
			"locations_set": {
				Type:     schema.TypeSet,
				MinItems: 1,
				Optional: true,
				Elem: &schema.Schema{
					Type: schema.TypeString,
					ValidateFunc: validation.NoZeroValues, //todo at some point we should write a location validation function..
				},
				Set: func(v interface{}) int {
					return hashcode.String(azureRMNormalizeLocation(v.(string)))
				},
			},
```

The tests now passes. I guess those functions were doing something weird to the state. The set function is just the `HashString` with a call to normalize location. I *think* that should work as intended now. I would recommend pulling that out into a `HashAzureLocation()` function.

jcorioland commented:
Hi @katbyte - Thank you very much for your support! I was able to move back on Set for the geo-replication locations. Everything is working fine now.
I have also added sample code for container-registry and updated the docs.

LMK if I need to update anything else in order to get this PR merge.

Thanks a lot!

jcorioland commented on azurerm/resource_arm_container_registry_test.go:
```diff
@@ -252,6 +254,82 @@ func TestAccAzureRMContainerRegistry_update(t *testing.T) {
+					resource.TestCheckResourceAttr(dataSourceName, "name", containerRegistryName),
+					resource.TestCheckResourceAttr(dataSourceName, "resource_group_name", resourceGroupName),
+					resource.TestCheckResourceAttr(dataSourceName, "sku", skuPremium),
+					testCheckAzureRMContainerRegistryExists(dataSourceName),
```
Done.

jcorioland commented on azurerm/resource_arm_container_registry_test.go:
```diff
@@ -252,6 +254,82 @@ func TestAccAzureRMContainerRegistry_update(t *testing.T) {
+					resource.TestCheckResourceAttr(dataSourceName, "resource_group_name", resourceGroupName),
+					resource.TestCheckResourceAttr(dataSourceName, "sku", skuPremium),
+					testCheckAzureRMContainerRegistryExists(dataSourceName),
+					testCheckAzureRMContainerRegistryGeoreplications(dataSourceName, skuPremium, []string{"\"eastus\"", "\"westus\""}),
```
Done.

jcorioland commented on examples/container-registry/variables.tf:
```diff
@@ -0,0 +1,28 @@
+  type        = "list"
+  description = "Azure locations where the Container Registry should be geo-replicated - Only with Premium SKU"
+  default     = ["eastus"]
+}
```
Done.

jcorioland commented on website/docs/r/container_registry.html.markdown:
```diff
@@ -40,6 +44,26 @@ resource "azurerm_container_registry" "test" {
 
+### Managed Container Registry
+
+When using another SKU than `Classic`, Azure Container Registry manages the storage account for you.
```
Done.

jcorioland commented on azurerm/resource_arm_container_registry_test.go:
```diff
@@ -252,6 +254,82 @@ func TestAccAzureRMContainerRegistry_update(t *testing.T) {
+	ri := acctest.RandInt()
+	containerRegistryName := fmt.Sprintf("testacccr%d", ri)
+	resourceGroupName := fmt.Sprintf("testAccRg-%d", ri)
+	config := testAccAzureRMContainerRegistry_geoReplication(ri, testLocation(), skuPremium, "\"eastus\", \"westus\"")
```
Done.

jcorioland commented on azurerm/resource_arm_container_registry_test.go:
```diff
@@ -252,6 +254,82 @@ func TestAccAzureRMContainerRegistry_update(t *testing.T) {
+	containerRegistryName := fmt.Sprintf("testacccr%d", ri)
+	resourceGroupName := fmt.Sprintf("testAccRg-%d", ri)
+	config := testAccAzureRMContainerRegistry_geoReplication(ri, testLocation(), skuPremium, "\"eastus\", \"westus\"")
+	updatedConfig := testAccAzureRMContainerRegistry_geoReplication(ri, testLocation(), skuPremium, "\"westeurope\", \"eastus\"")
```
Done.

jcorioland commented on azurerm/resource_arm_container_registry_test.go:
```diff
@@ -252,6 +254,82 @@ func TestAccAzureRMContainerRegistry_update(t *testing.T) {
+					resource.TestCheckResourceAttr(dataSourceName, "resource_group_name", resourceGroupName),
+					resource.TestCheckResourceAttr(dataSourceName, "sku", skuPremium),
+					testCheckAzureRMContainerRegistryExists(dataSourceName),
+					testCheckAzureRMContainerRegistryGeoreplications(dataSourceName, skuPremium, []string{"\"eastus\"", "\"westus\""}),
```
Done.

jcorioland commented on azurerm/resource_arm_container_registry_test.go:
```diff
@@ -252,6 +254,82 @@ func TestAccAzureRMContainerRegistry_update(t *testing.T) {
+					resource.TestCheckResourceAttr(dataSourceName, "resource_group_name", resourceGroupName),
+					resource.TestCheckResourceAttr(dataSourceName, "sku", skuPremium),
+					testCheckAzureRMContainerRegistryExists(dataSourceName),
+					testCheckAzureRMContainerRegistryGeoreplications(dataSourceName, skuPremium, []string{"\"eastus\"", "\"westeurope\""}),
```
Done.

jcorioland commented:
@katbyte thanks for the review. I've done the updates and validated that all unit tests pass for ContainerRegistry.

katbyte commented:
Thanks for the updates @jcorioland! tests look good now:
<img width="512" alt="screen shot 2018-11-27 at 10 42 11" src="https://user-images.githubusercontent.com/1638467/49103799-2bfb2a00-f231-11e8-886c-8877c9529dfe.png">
